homeassistant/components/modem_callerid/__init__.py

```diff
@@ -1 +1,72 @@
-"""The modem_callerid component."""
+"""The Modem Caller ID integration."""
+import asyncio
+import logging
+
+from phone_modem import PhoneModem, exceptions
+import serial
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
+from homeassistant.const import CONF_DEVICE
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import DATA_KEY_API, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = [SENSOR_DOMAIN]
+
+
+async def async_setup(hass: HomeAssistant, config):
+    """Set up the Modem Caller ID component."""
+    hass.data[DOMAIN] = {}
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config.get(DOMAIN)
+        )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up Modem Caller ID from a config entry."""
+    device = entry.data[CONF_DEVICE]
+    try:
+        api = PhoneModem(device)
+    except (
+        FileNotFoundError,
+        exceptions.SerialError,
+        serial.SerialException,
+        serial.serialutil.SerialException,
+    ) as ex:
+        _LOGGER.error("Unable to open port %s", device)
+        raise ConfigEntryNotReady from ex
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        DATA_KEY_API: api,
+    }
+
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
```

Comment:
Be sure to stop the thread in `phone-modem` when the config entry is unloaded.

Reply:
See `__init__.py:72`

Reply:
https://github.com/tkdrob/phone_modem/blob/master/phone_modem/__init__.py#L61

Store a reference to the thread in your library so it can be .join()ed on config entry unload (https://docs.python.org/3/library/threading.html#threading.Thread.join)

You've marked the thread as a daemon so it will be abruptly killed at shutdown (https://docs.python.org/3/library/threading.html#thread-objects) but not on unload.  You should use something like a threading.Event to tell the thread its time to exit (https://docs.python.org/3/library/threading.html#threading.Event) when you call a shutdown method.